crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2485,26 +2532,41 @@ impl<'db> TypeInferenceBuilder<'db> {
                         unreachable!("Not operator is handled in its own case");
                     }
                 };
-                let class_member = class.class_member(self.db, unary_dunder_method);
-                let call = class_member.call(self.db, &[operand_type]);
 
-                match call.return_ty_result(
-                    self.db,
-                    AnyNodeRef::ExprUnaryOp(unary),
-                    &mut self.diagnostics,
-                ) {
-                    Ok(t) => t,
-                    Err(e) => {
-                        self.diagnostics.add(
-                            unary.into(),
-                            "unsupported-operator",
-                            format_args!(
-                                "Unary operator `{op}` is unsupported for type `{}`",
-                                operand_type.display(self.db),
-                            ),
-                        );
-                        e.return_ty()
+                if let SymbolLookupResult::Type(class_member, _) =
+                    class.class_member(self.db, unary_dunder_method)
+                {
+                    let call = class_member.call(self.db, &[operand_type]);
+
+                    match call.return_ty_result(
+                        self.db,
+                        AnyNodeRef::ExprUnaryOp(unary),
+                        &mut self.diagnostics,
+                    ) {
+                        Ok(t) => t,
+                        Err(e) => {
+                            self.diagnostics.add(
+                                unary.into(),
+                                "unsupported-operator",
+                                format_args!(
+                                    "Unary operator `{op}` is unsupported for type `{}`",
+                                    operand_type.display(self.db),
+                                ),
+                            );
+                            e.return_ty()
+                        }
                     }
+                } else {
+                    self.diagnostics.add(
+                        unary.into(),
+                        "unsupported-operator",
+                        format_args!(
+                            "Unary operator `{op}` is unsupported for type `{}`",
+                            operand_type.display(self.db),
+                        ),
+                    );
+
+                    Type::Unknown
```

Comment:
It does seem like we may want to find a way to improve the ergonomics of this case, so we don't have to repeat the diagnostic? But we can also make repeating the diagnostic more ergonomic by making it a method on `self.diagnostics`.

Reply:
While I did that, I also found a new case that was not handled correctly. Added a test case:
```py
def bool_instance() -> bool:
    return True

if bool_instance():
    x = "abc"

class C:
    if bool_instance():
        x = 1

    # error: [possibly-unresolved-reference]
    y = x

reveal_type(C.y)  # revealed: Literal[1] | Literal["abc"]
```

Reply:
I don't see this new test case in the PR?

Reply:
Oops, never mind, I see it now; not sure how I missed it.